cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/AbstractServiceRetryableMacroEvaluator.java

```diff
@@ -106,10 +106,25 @@ protected String validateAndRetrieveContent(String serviceName,
     if (urlConn == null) {
       throw new RetryableException(serviceName + " service is not available");
     }
+    retrieveContent(serviceName, urlConn);
+    try (Reader reader = new InputStreamReader(urlConn.getInputStream(), StandardCharsets.UTF_8)) {
+      return CharStreams.toString(reader);
+    } finally {
+      urlConn.disconnect();
+    }
+  }
 
-    int responseCode = urlConn.getResponseCode();
+  private void retrieveContent(String serviceName, HttpURLConnection urlConn) throws IOException {
+    int responseCode;
+    try {
+      responseCode = urlConn.getResponseCode();
+    } catch (IOException e) {
+      // IOExceptions are retryable for idempotent operations.
+      throw new RetryableException(e);
+    }
     if (responseCode != HttpURLConnection.HTTP_OK) {
       switch (responseCode) {
+        case HttpURLConnection.HTTP_INTERNAL_ERROR:
```

Comment:
Can we start unifying approach with io.cdap.cdap.common.internal.remote.RemoteClient#executeIdempotent? Like have a common method that decides if given code is retryable or common list of codes. This way we don't need to update it in multiple places

Reply:
moved it to a list in `cdap-api-common/Constants`